bin/send_email.py

```diff
@@ -6,60 +6,152 @@
+                "0",
+            ): f"Submission {submission_id} has been evaluated. Your score will be available after Challenge submissions are closed. Thank you for participating!",
+            (
+                "INVALID",
+                "1",
+            ): f"Evaluation failed for Submission {submission_id}."
+            + "\n"
+            + f"Reason: '{reason}'."
+            + "\n"
+            + f"View your submissions here: https://www.synapse.org/#!Synapse:{view_id}/tables/, and contact the organizers for more information.",
+            (
+                "INVALID",
+                "0",
+            ): f"Evaluation failed for Submission {submission_id}."
+            + "\n"
+            + f"Reason: '{reason}'."
+            + "\n"
+            + "Please contact the organizers for more information.",
+        }
+
+        body = templates.get((status, email_with_score))
+
+        return body
+
+    def get_annotations(self, submission_id: str):
+        """
+        Gets the ``status`` ``score`` and ``reason`` annotations for the given
+        submission on Synapse and makes them attributes of the class instance.
+
+        1. ``status`` is the submission status, as defined by the last begun stage
+        in the MODEL_TO_SYNAPSE workflow.
+        2. ``score`` is the score of the model, used to determine its accuracy.
+        3. ``reason`` is the reason for the validation error, if there was one.
+        It remains an empty string (None) if no validation error.
+        """
+        submission_annotations = self.syn.getSubmissionStatus(submission_id)[
+            "submissionAnnotations"
+        ]
+        # TODO: "auc" may not always be the annotation name for score.
+        # Should enforce annotation names in the score/validation scripts.
+        self.status = submission_annotations.get("validation_status")[0]
+        self.score = submission_annotations.get("auc")[0]
+        self.reason = submission_annotations.get("validation_errors")[0]
+
+    def send_email(self, view_id: str, submission_id: str, email_with_score: str):
+        """
+        Sends an e-mail on the status of the individual submission
+        to the submitting team or individual.
+
+        Arguments:
+          view_id: The view Id of the Submission View on Synapse
+          submission_id: The ID for an individual submission within an evaluation queue
+
+        """
+        # Get MODEL_TO_DATA annotations for the given submission
+        self.get_annotations(submission_id)
+
+        # Get the Synapse users to send an e-mail to
+        ids_to_notify = self.get_participant_id(submission_id)
+
+        # Create the subject and body of the e-mail message, depending on submission status
+        subject = (
+            f"Evaluation Success: {submission_id}"
+            if self.status == "VALIDATED"
+            else f"Evaluation Failed: {submission_id}"
+        )
+        body = self.email_template(
+            self.status,
```

Comment:
nit: If `email_template` is only expected to be used after a class instance has been created and the `status`, `score` and `reason` attributes have been filled in there isn't a need to pass them to a method on self. You will already have a reference to these attributes through `self.{attribute}`

Reply:
I left `self.syn` out for that reason, but left `status` `score` and `reason` in because I felt that otherwise it would be unclear what information the function needs to generate its output. I suppose that is one of the signs telling me this should be better left as a function and not a method!